scripts/tf_cnn_benchmarks/benchmark_cnn_kungfu.py

```diff
@@ -2371,148 +2610,160 @@ def _benchmark_graph(self, graph_info, eval_graph_info):
+                and self.params.train_dir
+                and self.params.save_summaries_steps > 0):
+            summary_writer = tf.summary.FileWriter(self.params.train_dir,
+                                                   tf.get_default_graph())
+
+        # We want to start the benchmark timer right after a image_producer barrier
+        # and avoids undesired waiting times on barriers.
+        if ((self.num_warmup_batches + len(graph_info.enqueue_ops) - 1) %
+                self.batch_group_size) != 0:
+            self.num_warmup_batches = int(
+                math.ceil((self.num_warmup_batches +
+                           len(graph_info.enqueue_ops) - 1.0) /
+                          (self.batch_group_size)) * self.batch_group_size -
+                len(graph_info.enqueue_ops) + 1)
+            log_fn('Round up warm up steps to %d to match batch_group_size' %
+                   self.num_warmup_batches)
+            assert (
+                (self.num_warmup_batches + len(graph_info.enqueue_ops) - 1) %
+                self.batch_group_size) == 0
+        # We run the summaries in the same thread as the training operations by
+        # passing in None for summary_op to avoid a summary_thread being started.
+        # Running summaries and training operations in parallel could run out of
+        # GPU memory.
+        if is_chief and not self.forward_only_and_freeze:
+            saver = tf.train.Saver(self.variable_mgr.savable_variables(),
+                                   save_relative_paths=True,
+                                   max_to_keep=self.params.max_ckpts_to_keep)
+
+        if self.params.variable_update == 'kungfu' and not is_chief:
+            saver = tf.train.Saver(self.variable_mgr.savable_variables(),
+                                   save_relative_paths=True,
+                                   max_to_keep=self.params.max_ckpts_to_keep)
+
+        ready_for_local_init_op = None
+        if self.job_name and not (self.single_session
+                                  or self.distributed_collective):
+            # In distributed mode, we don't want to run local_var_init_op_group until
+            # the global variables are initialized, because local_var_init_op_group
+            # may use global variables (such as in distributed replicated mode). We
+            # don't set this in non-distributed mode, because in non-distributed mode,
+            # local_var_init_op_group may itself initialize global variables (such as
+            # in replicated mode).
+            ready_for_local_init_op = tf.report_uninitialized_variables(
+                tf.global_variables())
+        if self.params.variable_update == 'horovod':
+            import horovod.tensorflow as hvd  # pylint: disable=g-import-not-at-top
+            bcast_global_variables_op = hvd.broadcast_global_variables(0)
+        elif self.params.variable_update == 'kungfu':
+            import kungfu as kf
```

Comment:
@luomai @lgarithm The adaptive model averaging operator introduces a new TF boolean tensor:
```
   init_mask = tf.constant([r != rank for r in range(np)])
   neighbour_mask = tf.Variable(init_mask, trainable=False)
```

I created a new if branch to handle the parallel-sgd case because the `broadcast` operator for global variables does not support boolean tensors. Perhaps we can add support for bool as well.

Reply:
We could add support for broadcast bool, but in this case, `neighbour_mask` shouldn't be broadcasted. Currently it's already marked as `trainable=False`, we need to find out why it is still been broadcasted.